Add a MongoDB client extension

cescoffier opened the pull request:
This is the first version of the Mongo extension.

It:

* works in both JVM and native mode
* proposes an imperative client (regular mongo client) and a reactive one (based on the reactive streams mongo driver)
* supports authentication and TLS (even in native mode), but does not support `mongo+srv://` in native mode (SVM limitation)
* supports the hot reload

cescoffier committed:
First version of the mongodb extension

cescoffier committed:
Add a simple support for codec and codec provider

cescoffier committed:
Initial support of the MongoClient configuration.

cescoffier committed:
Provides a reactive API and an implementation. Improve configuration and feature coverage in native mode.

Inject both reactive and imperative client
Extend configuration support for both the imperative and reactive clients
Add an integration test for the reactive client
Fix credential and write concern configuration
Remove the new for having quarkus-reactive-streams-operators as a dependency of projects using the mongo client
Re-inject the client so it get the new config
Close the client during the hot reload so we don't have a connection shortage
Register SSL support if not explicitly disabled
Provide a substitution for the DNS lookup. However it won't work until Substrate provide a non-blocking multicast implementation.
Declare the mongo version and the mongo extension in the bom.

cescoffier committed:
Fix read preferences and other settings
Fix integration tests

It works better when you start the database, weird, isn't it?
Also align with the embedded mongo used in the runtime tests

cescoffier committed:
Explicitly mark the mongo+srv unsupported in native mode

Also, the number of databases created by default differs on Mac and Linux. So, reduce the number of documents as on Windows the number of connection is limited to 500.

cescoffier committed:
Add Mongo documentation

Co-Authored-By: Guillaume Smet <guillaume.smet@gmail.com>

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+
+If you need more configuration properties, there is a full list at the end of this guide.
+
+== Running a Mongo Database
```
```suggestion
== Running a MongoDB Database
```

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+If you need more configuration properties, there is a full list at the end of this guide.
+
+== Running a Mongo Database
+As by default, `MongoClient` is configure to access a local Mongo database on port 27017 (the default Mongo port), if you have a local running database on this port, there is nothing more to do before being able to test it!
```
```suggestion
As by default, `MongoClient` is configure to access a local MongoDB database on port 27017 (the default MongoDB port), if you have a local running database on this port, there is nothing more to do before being able to test it!
```

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+== Running a Mongo Database
+As by default, `MongoClient` is configure to access a local Mongo database on port 27017 (the default Mongo port), if you have a local running database on this port, there is nothing more to do before being able to test it!
+
+If you want to use Docker to run a Mongo database, you can use the following command to launch one:
```
```suggestion
If you want to use Docker to run a MongoDB database, you can use the following command to launch one:
```

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+
+Now let's add a simple web page to interact with our `FruitResource`.
+Quarkus automatically serves static resources located under the `META-INF/resources` directory.
+In the `src/main/resources/META-INF/resources` directory, add a `fruits.html` file with the content from this {quickstarts-blob-url}/rest-json/src/main/resources/META-INF/resources/fruits.html[fruits.html] file in it.
```
I don't think you're referring to the right quickstart?

cescoffier replied:
it's copying the same file, but yes, would be better to provide the file in the right quickstart. Keeping this comment open until we have the right url.

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+ * open a browser to `http://localhost:8080/fruits.html`
+ * add new fruits to the list via the form
+
+== Reactive Mongo Client
```
```suggestion
== Reactive MongoDB Client
```

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+ * add new fruits to the list via the form
+
+== Reactive Mongo Client
+A reactive Mongo Client is included in Quarkus.
```
```suggestion
A reactive MongoDB Client is included in Quarkus.
```

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+
+== Reactive Mongo Client
+A reactive Mongo Client is included in Quarkus.
+Using it is as easy as using the classic Mongo Client. You can rewrite the previous example to use it like the following.
```
```suggestion
Using it is as easy as using the classic MongoDB Client. You can rewrite the previous example to use it like the following.
```

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+
+== Simplifying Mongo Client usage
+One of the drawback of using directly the Mongo Client, is that you need to manually transform your objects to/from a Mongo `Document`.
+To facilitate this there is two ways.
```
```suggestion
To facilitate this there are two ways.
```
?

cescoffier replied:
section dropped.

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+}
+----
+
+== Simplifying Mongo Client usage
```
Same comment about MongoDB but I suppose it depends on how you want to name the extension in the end.

cescoffier replied:
Here as it's not followed with database, it should be MongoDB.

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+To facilitate this there is two ways.
+
+=== Using a Bson Codec
+By using a Bson `Codec`, the Mongo Client will take care of the transformation of your domain object to/from a Mongo `Document` automatically.
```
```suggestion
By using a Bson `Codec`, the Mongo Client will take care of the transformation of your domain object to/from a MongoDB `Document` automatically.
```

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+=== Using a Bson Codec
+By using a Bson `Codec`, the Mongo Client will take care of the transformation of your domain object to/from a Mongo `Document` automatically.
+
+First you need to create a Bson `Codec` that will tells Bson how to transform your entity to/from a Mongo `Document`.
```
```suggestion
First you need to create a Bson `Codec` that will tell Bson how to transform your entity to/from a MongoDB `Document`.
```

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+By using a Bson `Codec`, the Mongo Client will take care of the transformation of your domain object to/from a Mongo `Document` automatically.
+
+First you need to create a Bson `Codec` that will tells Bson how to transform your entity to/from a Mongo `Document`.
+Here we use a `CollectibleCodec` as our object is retrievable from the database (it has a Mongo identifier), if not we would have used a `Codec` instead.
```
```suggestion
Here we use a `CollectibleCodec` as our object is retrievable from the database (it has a MongoDB identifier), if not we would have used a `Codec` instead.
```

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+
+First you need to create a Bson `Codec` that will tells Bson how to transform your entity to/from a Mongo `Document`.
+Here we use a `CollectibleCodec` as our object is retrievable from the database (it has a Mongo identifier), if not we would have used a `Codec` instead.
+More information on the codec documentation : https://mongodb.github.io/mongo-java-driver/3.10/bson/codecs.
```
```suggestion
More information in the codec documentation: https://mongodb.github.io/mongo-java-driver/3.10/bson/codecs.
```

cescoffier replied:
fixed

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+----
+
+
+Then you need to create a `CodecProvider` to linked this `Codec` to the `Fruit` class.
```
```suggestion
Then you need to create a `CodecProvider` to link this `Codec` to the `Fruit` class.
```

cescoffier replied:
fixed.

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+
+== Conclusion
+
+Accessing a Mongo database from a Mongo Client is easy with Quarkus as it provides configuration and native support for it.
```
```suggestion
Accessing a MongoDB database from a Mongo Client is easy with Quarkus as it provides configuration and native support for it.
```

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+MongoDB connection string, this is a URI where a lot of options can be passed : https://docs.mongodb.com/manual/reference/connection-string/
+
+`quarkus.mongodb.hosts`::
+Coma separated lits of hosts.
```
```suggestion
Comma separated list of hosts.
```

cescoffier replied:
fixed.

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+Maximum number of connections in the connection pool.
+
+`quarkus.mongodb.min-pool-size`::
+Minumum number of connections in the connection pool.
```
```suggestion
Minimum number of connections in the connection pool.
```

cescoffier replied:
Also had to be fixed in the javadoc.

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+Minumum number of connections in the connection pool.
+
+`quarkus.mongodb.max-connection-idle-time`::
+Maximum idle time in milliseconds of a pooled connection. A connection that exceeds this limit will be closed.
```
Should they be `Duration` instead? Same question for all these variables.

(we have a note to include if this is the case)

cescoffier replied:
Wondering about it. This is the MongClient configuration passed almost directly to the configuration. But yes, I would prefer using duration.

cescoffier replied:
Fixed - used duration everywhere.

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+Configures period of time in milliseconds to wait before running the first maintenance job on the connection pool.
+
+`quarkus.mongodb.wait-queue-multiple`::
+This multiplier, multiplied with the `maxPoolSi` setting, gives the maximum number of threads that may be waiting for a connection to become available from the pool. All further threads will get an exception right away.
```
maxPoolSi?

cescoffier replied:
ze :-D

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+How long in milliseconds a send or receive on a socket can take before timing out.
+
+`quarkus.mongodb.tls-insecure`::
+If connecting with TLS, this option enables insecure TLS connections. Default to false.
```
```suggestion
If connecting with TLS, this option enables insecure TLS connections. Defaults to false.
```

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+If connecting with TLS, this option enables insecure TLS connections. Default to false.
+
+`quarkus.mongodb.tls`::
+Whether to connect using TLS. Default to false.
```
```suggestion
Whether to connect using TLS. Defaults to false.
```

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+`quarkus.mongodb.credentials.auth-mechanism`::
+Configures the authentication mechanism to use if a credential was supplied.
+The default is unspecified, in which case the client will pick the most secure mechanism available based on the
+sever version. For the GSSAPI and MONGODB-X509 mechanisms, no password is accepted, only the username.
```
```suggestion
server version. For the GSSAPI and MONGODB-X509 mechanisms, no password is accepted, only the username.
```

gsmet commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+
+`quarkus.mongodb.credentials.auth-source`::
+Configures the source of the authentication credentials.
+This is typically the database that the credentials have been created. The value defaults to the database
```
`the database that the credentials have been created` hmmm, `that` is weird here.

cescoffier replied:
Updated. This was actually in the upstream code...

gsmet reviewed (changes requested):
I made a big review of the documentation part, NOT the code.

I think we should try to be consistent about the Mongo*DB* naming but I let you decide what you prefer.

cescoffier committed:
Rename modules to quarkus-mongodb-client

This includes:
* module rename (artifactId, name...)
* package rename
* directory rename

Use duration instead of time in milliseconds.

Rename MongoDB directories, artifactIds, project name and package name

Replace SLF4J with JBoss Logging

cescoffier commented on core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java:
```diff
@@ -38,6 +38,7 @@
     public static final String KUBERNETES = "kubernetes";
     public static final String KUBERNETES_CLIENT = "kubernetes-client";
     public static final String MAILER = "mailer";
+    public static final String MONGO = "mongo-client";
```
Agreed!

cescoffier commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+
+Nothing fancy. One important thing to note is that having a default constructor is required by the JSON serialization layer.
+
+Now create a `org.acme.rest.json.FruitService` that will be the business layer of our application and store/load the fruits from the mongo database.
```
Here we cannot say MongoDB database, as DB is for database.

cescoffier commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+
+The implementation is pretty straightforward and you just need to define your endpoints using the JAX-RS annotations and use the FruitService to list/add new fruits.
+
+== Configuring the mongo database
```
Same comment as above, when followed with database, it should be mongo.

cescoffier commented on docs/src/main/asciidoc/mongo-guide.adoc:
```diff
@@ -0,0 +1,599 @@
+The implementation is pretty straightforward and you just need to define your endpoints using the JAX-RS annotations and use the FruitService to list/add new fruits.
+
+== Configuring the mongo database
+The main property to configure is the URL to access to MongoDB, almost all Mongo configuration can be put on the connection URI so we advise you to do so, you can find more information on the Mongo documentation site : https://docs.mongodb.com/manual/reference/connection-string/
```
Sentence rewritten.

cescoffier commented:
Ok, I believe we reached a milestone. I'm going to re-organized my commits and ask for a review.

gsmet commented on extensions/mongo-client/deployment/pom.xml:
```diff
@@ -0,0 +1,61 @@
+    </parent>
+
+    <artifactId>quarkus-mongodb-client-deployment</artifactId>
+    <name>Quarkus - Mongo - Client - Deployment</name>
```
```suggestion
    <name>Quarkus - MongoDB Client - Deployment</name>
```
would be better, I think.

cescoffier replied:
I changed the artifactId, forgot about the name. So definitely!

gsmet commented on extensions/mongo-client/pom.xml:
```diff
@@ -0,0 +1,21 @@
+
+   <modelVersion>4.0.0</modelVersion>
+   <artifactId>quarkus-mongodb-client-parent</artifactId>
+   <name>Quarkus - Mongo - Client</name>
```
```suggestion
   <name>Quarkus - MongoDB Client</name>
```

gsmet commented on extensions/mongo-client/runtime/pom.xml:
```diff
@@ -0,0 +1,89 @@
+    </parent>
+
+    <artifactId>quarkus-mongodb-client</artifactId>
+    <name>Quarkus - Mongo - Client - Runtime</name>
```
```suggestion
    <name>Quarkus - MongoDB Client - Runtime</name>
```

gsmet commented on extensions/pom.xml:
```diff
@@ -67,6 +67,7 @@
         <module>elasticsearch-rest-client</module>
         <module>kafka-client</module>
         <module>kafka-streams</module>
+        <module>mongo-client</module>
```
It should be mongodb-client for consistency.

gsmet commented on integration-tests/pom.xml:
```diff
@@ -52,6 +52,7 @@
         <module>amazon-lambda</module>
         <module>kogito</module>
         <module>kubernetes-client</module>
+        <module>mongo-client</module>
```
It should be mongodb-client for consistency.

gsmet commented on extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/CodecProviderBuildItem.java:
```diff
@@ -0,0 +1,18 @@
+package io.quarkus.mongo.deployment;
```
I think I would have made the package `io.quarkus.mongodb.client` (same for the ITs).

But I can understand how it's not practical to change it now. So feel free to ignore.

cescoffier replied:
It needs to be done. I actually started, but forgot to commit and pushed. It will be done today.

cescoffier replied:
Done.

gsmet commented on extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/MongoClientProcessor.java:
```diff
@@ -0,0 +1,73 @@
+    CodecProviderBuildItem collectCodecProviders(CombinedIndexBuildItem indexBuildItem) {
+        Collection<ClassInfo> codecProviderClasses = indexBuildItem.getIndex()
+                .getAllKnownImplementors(DotName.createSimple(CodecProvider.class.getName()));
+        List<String> names = codecProviderClasses.stream().map(ClassInfo::toString).collect(Collectors.toList());
```
Not excited about using the `toString()`. People sometimes change that without thinking that it could be used.

gsmet commented on extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/MongoClientProcessor.java:
```diff
@@ -0,0 +1,73 @@
+            sslNativeSupport.produce(new ExtensionSslNativeSupportBuildItem(FeatureBuildItem.MONGODB_CLIENT));
+        }
+
+        RuntimeValue<MongoClient> client = template.configureTheClient(config, beanContainer.getValue(),
```
Does your extension work if there's no config at all? We usually try to support that.

cescoffier replied:
That's a good question, let me try ;-)

cescoffier replied:
So, yes it works, and it connects to `mongodb://localhost:27017` (default host and port).

gsmet commented on extensions/mongo-client/runtime/src/main/java/io/quarkus/mongo/runtime/MongoClientTemplate.java:
```diff
@@ -0,0 +1,241 @@
+import org.bson.codecs.configuration.CodecRegistries;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
You don't want to use that. Please use JBoss Logging.

gsmet reviewed (changes requested):
I added another round of comments.

I still think there are too many commits. Typically https://github.com/quarkusio/quarkus/pull/3022/commits/ac046f26c2551472862f8f1db82b567b412e192c should be squashed as it really has no value.

gsmet commented on extensions/mongo-client/deployment/src/main/java/io/quarkus/mongo/deployment/MongoClientProcessor.java:
```diff
@@ -0,0 +1,73 @@
+
+        feature.produce(new FeatureBuildItem(FeatureBuildItem.MONGODB_CLIENT));
+
+        if (!sslNativeConfig.isExplicitlyDisabled()) {
```
You don't need to test that, it's automatically taken care of, just add the build item.

cescoffier committed:
Do not rely on ClassInfo.toString()

cescoffier committed:
Remove useless test checking whether SSL is explicitly disabled.

cescoffier commented:
I've reduced the number of commits.

cescoffier committed:
Update doc to use the new package name.

gsmet commented:
@cescoffier could you rebase, there are some conflicts with the pom files. It should be simple enough to fix.

gsmet commented on extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/impl/ReactiveMongoClientImpl.java:
```diff
@@ -0,0 +1,290 @@
+package io.quarkus.mongodb.impl;
```
The convention until now was to have the impl classes in the `runtime` package so it should be `io.quarkus.mongodb.runtime`.

cescoffier replied:
This module exposes an API, it would be weird to be in a `runtime` quarkus-specific structure. That being said, for implementation, I don't see an issue.

gsmet replied:
Well, this is the convention we decided. I wanted impl/spi packages but got ruled out.

So right now, the conventions we have are:
`io.quarkus.<extension>`: the potential runtime API
`io.quarkus.<extension>.runtime`: everything that is not API (you can have an `impl` subpackage here if you want)
`io.quarkus.<extension>.deployment`: everything from the deployment artifact

cescoffier commented:
Rebase done.

cescoffier committed:
Align the documentation with the quickstart

gsmet committed:
Rename template to recorder in the MongoDB extension

gsmet committed:
Add a missing @Override

gsmet commented:
I properly rebased to avoid the merge commit as it makes the history harder to read and adjusted a few things.

I will wait for CI and merge.

rsvoboda commented:
Thanks, just noticed that quarkus-quickstarts are broken due to missing quarkus-mongodb-client dependency for using-mongodb-client qs

gsmet commented:
@rsvoboda yes, I'm waiting for CI before merging this one.

gsmet commented:
And merged! Thanks @cescoffier and @loicmathieu !
